Change the trait bound `Arc<K>: Borrow<Q>` to `K: Borrow<Q>` (`sync` and `future` caches)

tatsuya6502 opened the pull request:
This PR changes the trait bound `Arc<K>: Borrow<Q>` to `K: Borrow<Q>` for the following methods of `sync` and `future` caches.

- `contains_key`
- `get`
- `invalidate`

This fixes #163 for `sync` and `future` caches. Now their methods will accept `&[u8]` as the key `&Q` when `K` is `Vec<u8>`.

JayKickliter commented:
Thanks for this. Just to be super clear: I used `K: Vec<u8>` and `Q: &[u8]` as my particular example, and wouldn't want to limit it to just that use case. That said, this PR seems to do the expected thing that Rust's std collections do.

JayKickliter commented:
This fixed my use-case https://github.com/JayKickliter/cream/pull/1/files#diff-28b0ccc836e6e883168b88ebb6f39508969ece5516e5f81458abd33592b12430

tatsuya6502 commented:
@JayKickliter

I found a better solution and I updated this branch. The old solution required the key of the internal hash table should be always wrapped in an `Arc`. But there is one hash table usage in `moka` that I did not want to wrap the key with `Arc`:

https://github.com/moka-rs/moka/pull/167/commits/848826852a8e3949499a5bc9fd31e8e7f6c31a20#diff-3df6f2ad1c01aeaaff49319246d71a65d7fd5ac8ae5bcd0c32a912ba38228f65L263-R263

```diff
-        (cht_key, hash)
+        (Arc::new(cht_key), hash)
```

(I wanted to revert the above change)

The new solution does not require it, so I reverted the above change and I am happy now.

I tried `cream`'s `jsk/bump-moka` branch with the new solution and it worked. (I am using Erlang/OTP at work, so I had it installed on my Mac)

```console
$ rustc -V           
rustc 1.62.0 (a8314ef7d 2022-06-27)

$ asdf current erlang
erlang          25.0.3          ... /cream/.tool-versions

$ git remote -v
origin	git@github.com:JayKickliter/cream.git (fetch)
origin	git@github.com:JayKickliter/cream.git (push)

$ git status
On branch bump-moka
Your branch is up to date with 'origin/jsk/bump-moka'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   Cargo.lock
	modified:   Cargo.toml

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	.tool-versions

no changes added to commit (use "git add" and/or "git commit -a")

$ git diff Cargo.toml 
diff --git a/Cargo.toml b/Cargo.toml
index 586dd68..7934353 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -10,6 +10,6 @@ name = "cream_nif"
 path = "native/lib.rs"
 
 [dependencies]
-moka = { git = "https://github.com/moka-rs/moka.git", rev = "dcae5d520553b16d49f25922a6b31442aff2a4fa" }
+moka = { git = "https://github.com/moka-rs/moka.git", rev = "865d0899b77f23a9de73accd2c1a4154ac59ffad" }
 rustler = "0.25.0"
 rustler_stored_term = "0.1.0"

$ cargo tree -i moka
moka v0.9.1 (https://github.com/moka-rs/moka.git?rev=865d0899b77f23a9de73accd2c1a4154ac59ffad#865d0899)
└── cream v0.1.0 (... /cream)

$ rebar3 shell    
===> Verifying dependencies...
...
===> Analyzing applications...
===> Compiling cream
Erlang/OTP 25 [erts-13.0.3] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [jit]

Eshell V13.0.3  (abort with ^G)
1> MaxCapacity = 3.
3

2> {ok, Cache} = cream:new(MaxCapacity).
{ok,#Ref<0.1419974579.4198105090.215813>}

3> ExpensiveOperation = fun() -> io:format("value is not in cache, performing expensive operation\n"), 1 end.
#Fun<erl_eval.43.3316493>

4> 1 = cream:cache(Cache, {my, key}, ExpensiveOperation).
value is not in cache, performing expensive operation
1

5> 1 = cream:cache(Cache, {my, key}, ExpensiveOperation).
1
```

tatsuya6502 reviewed (commented):
Merging.

JayKickliter commented:
thanks so much for the quick fix (and for testing my own code)!
